### PR TITLE
[Unification][FLAVA] Refactor FLAVA with TransformerEncoder

### PR DESCRIPTION
**Pull request by RdoubleA**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #253
* #249
* #248
* #244
* __->__ #243
* #240
* #239

## Summary
Replace's `FLAVATransformerEncoder` and `FLAVATransformerOutput` with unified classes. Refactors text and image encoders, and some builder functions.

FLAVA checks
- Manually loaded the checkpoint on main branch and this branch, outputs were the exact same
- Ran COCO zero shot eval script and got exact same recall numbers

## Test Plan
`pytest test/models/flava -vv`
```
=============================================== test session starts ================================================
platform linux -- Python 3.9.12, pytest-7.1.1, pluggy-1.0.0 -- /fsx/users/rafiayub/conda/envs/torchmm/bin/python
cachedir: .pytest_cache
rootdir: /data/home/rafiayub/torchmultimodal, configfile: pyproject.toml
plugins: hydra-core-1.1.2, cov-3.0.0, mock-3.8.2
collected 16 items                                                                                                 

test/models/flava/test_checkpoint.py::TestFLAVACheckpoint::test_flava_model_for_classification PASSED        [  6%]
test/models/flava/test_checkpoint.py::TestFLAVACheckpoint::test_flava_model_for_pretraining PASSED           [ 12%]
test/models/flava/test_checkpoint.py::TestFLAVACheckpoint::test_flava_model PASSED                           [ 18%]
test/models/flava/test_flava.py::TestFLAVA::test_forward_classification PASSED                               [ 25%]
test/models/flava/test_flava.py::TestFLAVA::test_forward_pretraining PASSED                                  [ 31%]
test/models/flava/test_flava.py::TestFLAVAModel::test_forward_image PASSED                                   [ 37%]
test/models/flava/test_flava.py::TestFLAVAModel::test_forward_image_text PASSED                              [ 43%]
test/models/flava/test_flava.py::TestFLAVAModel::test_forward_masked_image PASSED                            [ 50%]
test/models/flava/test_flava.py::TestFLAVAModel::test_forward_masked_image_and_text PASSED                   [ 56%]
test/models/flava/test_flava.py::TestFLAVAModel::test_forward_masked_text PASSED                             [ 62%]
test/models/flava/test_flava.py::TestFLAVAModel::test_forward_text PASSED                                    [ 68%]
test/models/flava/test_image_encoder.py::TestFlavaImageEncoder::test_embedding PASSED                        [ 75%]
test/models/flava/test_image_encoder.py::TestFlavaImageEncoder::test_image_encoder PASSED                    [ 81%]
test/models/flava/test_text_encoder.py::TestFlavaTextEncoder::test_embedding PASSED                          [ 87%]
test/models/flava/test_text_encoder.py::TestFlavaTextEncoder::test_text_transformer PASSED                   [ 93%]
test/models/flava/test_text_encoder.py::TestFlavaTextEncoder::test_text_transformer_attn_mask PASSED         [100%]

================================================= warnings summary =================================================
test/models/flava/test_checkpoint.py::TestFLAVACheckpoint::test_flava_model_for_pretraining
  /data/home/rafiayub/torchmultimodal/test/models/flava/test_checkpoint.py:102: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
    else torch.tensor(dict_actual[key])

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
========================================== 16 passed, 1 warning in 21.28s ==========================================
```
`python -m flava.coco_zero_shot --data_root /datasets01/COCO/022719/val2017 --annotations /datasets01/COCO/022719/annotations/captions_val2017.json`
```
INFO:root:image_to_text_recall@1 0.39660000801086426
INFO:root:image_to_text_recall@5 0.7110000252723694
INFO:root:text_to_image_recall@1 0.3959999978542328
INFO:root:text_to_image_recall@5 0.6958000063896179
```

Differential Revision: [D38506906](https://our.internmc.facebook.com/intern/diff/D38506906)

### CODE REVIEW

**Comment by RdoubleA** [comment]
@RdoubleA has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38506906).

**Comment by RdoubleA** [comment]
@RdoubleA has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38506906).

**Comment by RdoubleA** [comment]
@RdoubleA has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38506906).

**Comment by RdoubleA** [comment]
@RdoubleA has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38506906).

File: torchmultimodal/models/flava/model.py [hunk 1]
```diff
@@ -171,7 +170,7 @@ def forward(
                 text_encoding_out[1],
             )
         else:
-            text_outputs = text_encoding_out
+            text_outputs = text_encoding_out  # type: ignore
```
Comment (ankitade) [inline torchmultimodal/models/flava/model.py:173]: why ignore?

File: torchmultimodal/models/flava/model.py [hunk 2]
```diff
@@ -436,10 +432,9 @@ def flava_model(
     text_hidden_size: int = 768,
     text_num_attention_heads: int = 12,
     text_num_hidden_layers: int = 12,
-    text_hidden_dropout_prob: float = 0.0,
+    text_dropout: float = 0.0,
```
Comment (ankitade) [inline torchmultimodal/models/flava/model.py:435]: these names need not be changed right?

File: torchmultimodal/models/flava/text_encoder.py [hunk 3]
```diff
@@ -200,19 +201,18 @@ def flava_text_encoder(
...
-        intermediate_size=intermediate_size,
-        intermediate_activation=intermediate_activation,
-        attention_probs_dropout_prob=attention_probs_dropout_prob,
+    encoder = TransformerEncoder(
+        n_layer=num_hidden_layers,
+        d_model=hidden_size,
+        n_head=num_attention_heads,
+        dim_feedforward=intermediate_size,
+        activation=intermediate_activation,
         layer_norm_eps=layer_norm_eps,
```
Comment (ankitade) [inline torchmultimodal/models/flava/text_encoder.py:213]: where was pad_token_id used, now that we are removing it?

Comment (RdoubleA) [reply torchmultimodal/models/flava/text_encoder.py:213]: it was not used anywhere in `FLAVATransformerEncoder` or `FLAVATransformerLayer`

File: torchmultimodal/models/flava/model.py (hunk 2 above)
Comment (RdoubleA) [reply torchmultimodal/models/flava/model.py:435]: it was originally only for the dropout in the MLP in the encoder layer and MHA dropout was a separate argument, but I combined into one. FLAVA does not use either by default.

File: torchmultimodal/models/flava/model.py (hunk 1 above)
Comment (RdoubleA) [reply torchmultimodal/models/flava/model.py:173]: I was getting mypy type errors from these that I could not work around...

**Review by ankitade — APPROVED** [review]
